testcases/kernel/security/stack_clash/stack_clash.c

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2017 Pavel Boldin <pboldin@cloudlinux.com>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Adapted from code by Michal Hocko.
+ */
+
+/* This is a regression test of the Stack Clash [1] vulnerability.
+ * This tests that there is at least 1MiB of stack guard page which is
+ * considered hard to hop above. Code adapted from the Novell's bugzilla [2].
+ *
+ * [1] https://blog.qualys.com/securitylabs/2017/06/19/the-stack-clash
+ * [2] https://bugzilla.novell.com/show_bug.cgi?id=CVE-2017-1000364 */
+
+#include <sys/mman.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <alloca.h>
+#include <signal.h>
+#include <stdlib.h>
+
+#include "tst_test.h"
+
+static long PAGE_SIZE;
+static long PAGE_MASK;
+static int STACK_GROWSDOWN;
+
+#define SIGNAL_STACK_SIZE	(1UL<<20)
+#define MEGABYTE		(1<<20)
+#define THRESHOLD		(250 * PAGE_SIZE)
+#define FRAME_SIZE		1024
+
+void recurse(void)
+{
+	volatile char * ptr = alloca(FRAME_SIZE - sizeof(long));
+	*ptr = '\0';
+	recurse();
+}
+
+#define MAPPED_LEN PAGE_SIZE
+static unsigned long mapped_addr;
+void segv_handler(int sig, siginfo_t *info, void *data)
+{
+	unsigned long fault_addr = (unsigned long)info->si_addr;
+	unsigned long mmap_end = mapped_addr + MAPPED_LEN;
+	ssize_t diff;
+
+	if (sig != SIGSEGV)
+		return;
+
+	(void) data;
+
+	if (STACK_GROWSDOWN)
+		diff = fault_addr - mmap_end;
+	else
+		diff = mapped_addr - fault_addr;
+
+	tst_res(TINFO, "mmap = [%lx, %lx), addr = %lx, diff = %lx",
+		mapped_addr, mmap_end, fault_addr, diff);
+	if (diff < THRESHOLD)
+		exit(EXIT_FAILURE);
+	else
+		exit(EXIT_SUCCESS);
+}
+
+void do_child(void)
+{
+	int stack_top;
+	unsigned long stack_addr = (unsigned long)&stack_top;
+	stack_t signal_stack;
+	struct sigaction segv_sig = {.sa_sigaction = segv_handler, .sa_flags = SA_ONSTACK|SA_SIGINFO};
+	struct rlimit rlimit;
+
+	SAFE_GETRLIMIT(RLIMIT_STACK, &rlimit);
+	if (rlimit.rlim_cur == RLIM_INFINITY)
+		rlimit.rlim_cur = 16 * MEGABYTE;
```

Comment:
> Also default RLIMIT_STACK can smaller than 16M

Can you please elaborate?

As I see it in this case the code would simple use value provided by GETRLIMIT, and default stack size would be reflected there.